### ISSUE TITLE
Incorrect results with `prange` and sliced array

### ISSUE THREAD

**Issue by gmarkall** [numba/numba#7676]
This is a simplified version of the reproducer from https://numba.discourse.group/t/how-do-i-parallelize-this-code/1085/6?u=gmarkall.

The following:

```python
import numpy as np
from numba import prange, njit


@njit(parallel=True)
def myFoo(x, par, numthreads):
    n_par = par.shape[0]
    n_x = len(x)
    result = np.zeros((n_par, n_x), dtype=np.float64)
    chunklen = (len(x) + numthreads - 1) // numthreads

    for i in prange(numthreads):
        start = i * chunklen
        stop = (i + 1) * chunklen
        result[:, start:stop] = x[start:stop] * par[:, 0:1]

    return result


x = np.array(np.arange(0, 12, 1.0))
par = np.array([[1.0, 0.5], [2.0, 3.0], [3.0, 5.0]])
print(myFoo(x, par, 4))
print(myFoo.py_func(x, par, 4))
```

when run with `parallel=False` gives:

```
[[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11.]
 [ 0.  2.  4.  6.  8. 10. 12. 14. 16. 18. 20. 22.]
 [ 0.  3.  6.  9. 12. 15. 18. 21. 24. 27. 30. 33.]]
[[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11.]
 [ 0.  2.  4.  6.  8. 10. 12. 14. 16. 18. 20. 22.]
 [ 0.  3.  6.  9. 12. 15. 18. 21. 24. 27. 30. 33.]]
```

but when run with `parallel=True` gives:

```
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]
[[ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11.]
 [ 0.  2.  4.  6.  8. 10. 12. 14. 16. 18. 20. 22.]
 [ 0.  3.  6.  9. 12. 15. 18. 21. 24. 27. 30. 33.]]
```

Could you take a look please @DrTodd13 ?

**Comment by DrTodd13** [comment]
TL;DR Another parallel=True doesn't supporting broadcasting issue but in this case the parfor assertion is happening but not being given to the user.

@stuartarchibald @sklam Issue7675 is similar to this one.  In 7675, I solved it by seeing that a setitem had differing dimensions on the left and right hand sides and if so then disallowing that setitem from becoming a parfor.  In this case, the right-hand side (x[start:stop] * par[:, 0:1]) is a problem because of the dimension mismatch.  If you say that parfors won't support that (which is true) and you try to disallow it at this early stage because you see the dimensionality is different (versus a dynamic broadcasting check which you can only do at runtime) then you end up with an arrayexpr inside a parfor.  You might think that is okay but in the outlining  process, the code goes through type inference which doesn't know about arrayexpr.  I hacked something to get that to work and then onto the next problem which is that array analysis saw the arrayexpr and inserted an equivalence check even though now I made it so that a parfor wouldn't be created on a dimension mismatch.  The final problem I note here is that in my modified code (and I think in the original code) there is an assert_equiv that is firing and terminating the parfor gufunc but that error is not being propagated.  It looks to be eaten somewhere and not reported to the user so rather than at least some error they just get the wrong answer.

One final note before I forget.  Since we have types for what is in the parfor, a lot of the typing phase of the parfor gufunc compilation can be viewed as unneeded.  There are parts of the parfor gufunc that haven't been typed  so maybe we don't have a choice but to go through it again but in that other PR that is outstanding where we avoid going through redundant parts of the compilation path for the parfor gufunc we could consider if there's a way to speedup the gufunc compilation by reusing some of the work that's already done.

**Comment by gmarkall** [comment]
> TL;DR Another parallel=True doesn't supporting broadcasting issue but in this case the parfor assertion is happening but not being given to the user.

Slightly tangential, but do we need to go through the parfors code and replace its assertions with `NumbaError` instances? Would this help with the propagation of errors back to the user?
